lib/puppet/util/windows/registry.rb

```diff
@@ -271,14 +273,45 @@ def query_value_ex(key, name_ptr, &block)
                                     FFI::Pointer::NULL, type_ptr,
                                     FFI::Pointer::NULL, length_ptr)
 
-          FFI::MemoryPointer.new(:byte, length_ptr.read_dword) do |buffer_ptr|
+          # The call to RegQueryValueExW below is potentially unsafe:
+          # https://learn.microsoft.com/en-us/windows/win32/api/winreg/nf-winreg-regqueryvalueexw
+          #
+          #   "If the data has the REG_SZ, REG_MULTI_SZ or REG_EXPAND_SZ type,
+          #   the string may not have been stored with the proper terminating
+          #   null characters. Therefore, even if the function returns
+          #   ERROR_SUCCESS, the application should ensure that the string is
+          #   properly terminated before using it; otherwise, it may overwrite a
+          #   buffer. (Note that REG_MULTI_SZ strings should have two
+          #   terminating null characters.)"
+          #
+          # Since we don't know if the values will be properly null terminated,
+          # extend the buffer to guarantee we can append one or two wide null
+          # characters, without overwriting any data.
+          buf_bytes_len = case type_ptr.read_dword
+                          when Win32::Registry::REG_SZ, Win32::Registry::REG_EXPAND_SZ
+                            length_ptr.read_dword + WCHAR_SIZE
+                          when Win32::Registry::REG_MULTI_SZ
+                            length_ptr.read_dword + (WCHAR_SIZE * 2)
+                          else
+                            length_ptr.read_dword
+                          end
+
+          FFI::MemoryPointer.new(:byte, buf_bytes_len) do |buffer_ptr|
             result = RegQueryValueExW(key.hkey, name_ptr,
```

Comment:
We should probably switch to the safer [`RegGetValueW`](https://learn.microsoft.com/en-us/windows/win32/api/winreg/nf-winreg-reggetvaluew) to avoid this issue altogether.

Reply:
Going to punt on that and go with what we have for now